### ISSUE TITLE
Windows performance in Release profile seems crippled when building dev Cargo profile (RelWithDebInfo is faster)

### ISSUE THREAD

**Issue by vlovich** [utilityai/llama-cpp-rs#649]
I'm not sure if this is something specific to my configuration, but I'm seeing this across the board in CI & local machine: the Release variant compiler flags are missing O2 (& DNDEBUG). I don't think this has anything to do with this project per se as I suspect the bug lies in the cmake crate, but wanted to flag. This results in CPU inference being >2x slower on my AMD with performance equivalent to debug llama-cli. Switching to RelWithDebInfo speeds it up quite a bit.

```
//Flags used by the CXX compiler during RELEASE builds.
CMAKE_CXX_FLAGS_RELEASE:STRING= -nologo -MD -Brepro

//Flags used by the CXX compiler during RELWITHDEBINFO builds.
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG

//Flags used by the C compiler during RELEASE builds.
CMAKE_C_FLAGS_RELEASE:STRING= -nologo -MD -Brepro

//Flags used by the C compiler during RELWITHDEBINFO builds.
CMAKE_C_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG
```

It's also worth noting that stock cmake configures `/Ob2` as well for Release but I didn't observe any meaningful performance difference from that.

**Comment by babichjacob** [comment]
Yeah, I've observed slower-than-expected performance using this crate in comparison to the llama.cpp CLI (taking the avx2 binary from their GitHub releases).

**Comment by vlovich** [comment]
@babichjacob please share. I'm also observing a ~30% slowdown vs stock llama.cpp CLI. Is that ballpark what you're seeing as well? Would love to track that down together.

**Comment by babichjacob** [comment]
Seems even slower for me. I did a crude comparison with `hyperfine` and saw an approximately 6x slowdown. I'll report back with increasingly sophisticated measurements as I learn to extract them.

**Comment by AsbjornOlling** [comment]
We noticed the same thing in [NobodyWho](https://github.com/nobodywho-ooo/nobodywho) last week. Not sure if it's a new thing, or if we're only just noticing it.

We mostly work on linux, where we have like 98% the performance of llama-bench, but a windows user of our project noticed that it is significantly slower than llama.cpp

FWIW, we're seeing that it runs at like 23% of the t/s that pure llama.cpp does (only on windows).
![Image](https://github.com/user-attachments/assets/45b17915-962d-4586-b545-fec4b4755cfe)

I *just* finished setting up a windows VM to investigate 😅 Will post results if I find anything noteworthy.

**Comment by AsbjornOlling** [comment]
I'm not convinced that the windows performance reduction was fixed by #651 (or I don't understand it)

I'm slowly getting better at profiling and performance analysis, and haven't found anything super scary in the flamegraphs yet.

But a weird thing I'm noticing is that the numbers I get from binaries built with `cargo build --release` is significantly slower than the numbers I get from binaries built with just `cargo build`.

Gut feel is that something is off in `build.rs`. I'll have a closer look at it when I get the time

**Comment by vlovich** [comment]
@AsbjornOlling even in Rust debug builds, build.rs tried to build llama.cpp itself optimized. However, on Windows only, the underlying cmake crate has a bug where it explicitly turns off optimization flags for MSVC release. This ONLY impacts debug builds (ie cargo build) and not release (cargo build --release). After my patch I saw both debug and release had identical performance but that number was still 30% off from llama.cpp CLI. Please let me know if you have any insights.

**Comment by babichjacob** [comment]
I can concur that `LLAMA_LIB_PROFILE=RelWithDebInfo` and `LLAMA_LIB_PROFILE=MinSizeRel` both rocket the speed up (from 3.6 t/s ish to 10.7ish) to nearly match llama.cpp (11.2ish) for me.

**Comment by babichjacob** [comment]
(sorry about the comment right before this that I just deleted; I forgot to `pull` to actually test with the PR merged in 🤦)

I can corroborate what Asbjørn has found. Running on Windows 11 using MSVC with a Ryzen 5 5600 (so AVX2 is available):

| cargo profile | LLAMA_LIB_PROFILE | generation speed in my crude one-shot test is close to llama.cpp native |
|---------------|-------------------|-------------------------------------------------------------------------|
| release       | RelWithDebInfo    | ❌                                                                       |
| release       | MinSizeRel        | ❌                                                                       |
| release       | Release           | ❌                                                                       |
| release       | (unspecified)     | ❌                                                                       |
| debug         | RelWithDebInfo    | ❌                                                                       |
| debug         | MinSizeRel        | ❌                                                                       |
| debug         | Release           | ✅                                                                       |
| debug         | (unspecified)     | ✅                                                                       |

Where the ✅s are around 11 t/s and the ❌s are around 4 t/s for me.

The command I was getting measurements from was `cargo run --features native --example usage .../Qwen2.5-3B-Instruct-Q6_K_L.gguf` (with modifications to the command as reflected in the table). I ran `cargo clean` in between to prevent weirdness.

<details>
<summary>I modified <code>usage.rs</code> to get measurements like this:</summary>
<pre>
//! # Usage
//!
//! This is just about the smallest possible way to do inference. To fetch a model from hugging face:
//!
//! ```console
//! git clone --recursive https://github.com/utilityai/llama-cpp-rs
//! cd llama-cpp-rs/examples/usage
//! wget https://huggingface.co/Qwen/Qwen2-1.5B-Instruct-GGUF/resolve/main/qwen2-1_5b-instruct-q4_0.gguf
//! cargo run --example usage -- qwen2-1_5b-instruct-q4_0.gguf
//! ```
use llama_cpp_2::context::params::LlamaContextParams;
use llama_cpp_2::llama_backend::LlamaBackend;
use llama_cpp_2::llama_batch::LlamaBatch;
use llama_cpp_2::model::params::LlamaModelParams;
use llama_cpp_2::model::LlamaModel;
use llama_cpp_2::model::{AddBos, Special};
use llama_cpp_2::sampling::LlamaSampler;
use llama_cpp_2::{ggml_time_us, send_logs_to_tracing, LogOptions};
use std::io::Write;
use std::time::Duration;
</pre>
<pre>
#[allow(clippy::cast_possible_wrap, clippy::cast_possible_truncation)]
fn main() {
    let model_path = std::env::args().nth(1).expect("Please specify model path");

    tracing_subscriber::fmt::init();
    send_logs_to_tracing(LogOptions::default().with_logs_enabled(true));

    let backend = LlamaBackend::init().unwrap();
    let params = LlamaModelParams::default();

    let prompt =
        "<|im_start|>user\nWrite a 1000 word paragraph about llama-cpp-rs, a way to use llama.cpp from Rust.<|im_end|>\n<|im_start|>assistant\n".to_string();
    LlamaContextParams::default();
    let model =
        LlamaModel::load_from_file(&backend, model_path, &params).expect("unable to load model");
    let ctx_params = LlamaContextParams::default();
    let mut ctx = model
        .new_context(&backend, ctx_params)
        .expect("unable to create the llama_context");
    let tokens_list = model
        .str_to_token(&prompt, AddBos::Always)
        .unwrap_or_else(|_| panic!("failed to tokenize {prompt}"));
    let n_len = 128;

    // create a llama_batch with size 512
    // we use this object to submit token data for decoding
    let mut batch = LlamaBatch::new(512, 1);

    let last_index = tokens_list.len() as i32 - 1;
    for (i, token) in (0_i32..).zip(tokens_list.into_iter()) {
        // llama_decode will output logits only for the last token of the prompt
        let is_last = i == last_index;
        batch.add(token, i, &[0], is_last).unwrap();
    }
    ctx.decode(&mut batch).expect("llama_decode() failed");

    let mut n_cur = batch.n_tokens();
    let mut n_decode = 0;

    let t_main_start = ggml_time_us();

    // The `Decoder`
    let mut decoder = encoding_rs::UTF_8.new_decoder();
    let mut sampler = LlamaSampler::greedy();

    while n_decode < n_len {
        // sample the next token
        {
            let token = sampler.sample(&ctx, batch.n_tokens() - 1);

            sampler.accept(token);

            // is it an end of stream?
            if token == model.token_eos() {
                eprintln!();
                break;
            }

            let output_bytes = model.token_to_bytes(token, Special::Tokenize).unwrap();
            // use `Decoder.decode_to_string()` to avoid the intermediate buffer
            let mut output_string = String::with_capacity(32);
            let _decode_result = decoder.decode_to_string(&output_bytes, &mut output_string, false);
            print!("{output_string}");
            std::io::stdout().flush().unwrap();

            batch.clear();
            batch.add(token, n_cur, &[0], true).unwrap();
        }

        n_cur += 1;

        ctx.decode(&mut batch).expect("failed to eval");

        n_decode += 1;
    }
    eprintln!();

    let t_main_end = ggml_time_us();

    let duration = Duration::from_micros((t_main_end - t_main_start) as u64);

    eprintln!(
        "decoded {} tokens in {:.2} s, speed {:.2} t/s\n",
        n_decode,
        duration.as_secs_f32(),
        n_decode as f32 / duration.as_secs_f32()
    );
}
</pre>
</details>

**Comment by vlovich** [comment]
@babichjacob look at the generated CMakeCache.txt and the CXXFLAGS_RELEASE and CFLAGS_RELEASE variables. The /O2 is only missing when Rust is built debug. In --release the /O2 is there. I'm not sure what you're seeing but I linked the upstream bug report where I highlight the relevant code from cmake.rs which indeed only applies their opt flag filtering erroneously when building Rust debug but configuring cmake to be release.

**Comment by babichjacob** [comment]
Thanks for the hint. Really tricky situation.

> look at the generated CMakeCache.txt and the CXXFLAGS_RELEASE and CFLAGS_RELEASE variables. The /O2 is only missing when Rust is built debug. In --release the /O2 is there. 

Regardless of if your patch is enabled, and regardless of the cargo / Rust profile (where you're saying it should be there when compiled for release), `/O2` is missing for me in the `RELEASE` flags:

```
CMAKE_CXX_FLAGS:STRING= -nologo -MD -Brepro
CMAKE_CXX_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_CXX_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG
vvvv line of interest (optimization flags are missing especially in comparison to MINSIZERE and RELWITHDEBINFO)
CMAKE_CXX_FLAGS_RELEASE:STRING= -nologo -MD -Brepro
^^^^ line of interest
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG

CMAKE_C_FLAGS:STRING= -nologo -MD -Brepro
CMAKE_C_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_C_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG
vvvv line of interest (optimization flags are missing especially in comparison to MINSIZERE and RELWITHDEBINFO)
CMAKE_C_FLAGS_RELEASE:STRING= -nologo -MD -Brepro
^^^^ line of interest
CMAKE_C_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG
```

I suspect this is how it is for Asbjørn too. I'll also try wrangling a GitHub Actions worker and see what its flags look like.

With your patch enabled, I can see that `/O2` was indeed added (in the general flags section), but only in cargo's debug profile (because of the conditional mentioned above):

```
vvvv line of interest (optimization flags are the ones manually specified in build.rs)
CMAKE_CXX_FLAGS:STRING= /O2 /DNDEBUG /Ob2 -nologo -MD -Brepro 
^^^^ line of interest
CMAKE_CXX_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_CXX_FLAGS_MINSIZEREL:STRING=/MD /O1
CMAKE_CXX_FLAGS_RELEASE:STRING=
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG                                       

vvvv line of interest (optimization flags are the ones manually specified in build.rs)
CMAKE_C_FLAGS:STRING= /O2 /DNDEBUG /Ob2 -nologo -MD -Brepro
^^^^ line of interest
CMAKE_C_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1                   
CMAKE_C_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG            
CMAKE_C_FLAGS_RELEASE:STRING= /O2 /DNDEBUG /Ob2 -nologo
CMAKE_C_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG
```
(You can also see here that `cmake-rs` gave `/O2 /DNDEBUG /Ob2` to the C `RELEASE` flags but not C++. Weird!)

So, if I understand you correctly, this is what you see in release mode (regardless of if your patch is applied because it's conditional on debug):
```
CMAKE_CXX_FLAGS:STRING= -nologo -MD -Brepro
CMAKE_CXX_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_CXX_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG
vvvv line of interest (the expected optimization flags are present)
CMAKE_CXX_FLAGS_RELEASE:STRING=/O2 /DNDEBUG /Ob2 -nologo -MD -Brepro
^^^^ line of interest
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG

CMAKE_C_FLAGS:STRING= -nologo -MD -Brepro
CMAKE_C_FLAGS_DEBUG:STRING=/MDd /Zi /Ob0 /Od /RTC1
CMAKE_C_FLAGS_MINSIZEREL:STRING=/MD /O1 /Ob1 /DNDEBUG
vvvv line of interest (the expected optimization flags are present)
CMAKE_C_FLAGS_RELEASE:STRING=/O2 /DNDEBUG /Ob2 -nologo -MD -Brepro
^^^^ line of interest
CMAKE_C_FLAGS_RELWITHDEBINFO:STRING=/MD /Zi /O2 /Ob1 /DNDEBUG
```
Which is also the expected arrangement if the bug were resolved and the workaround wasn't needed (and it was regardless of whether cargo is in the release or debug profile).

I looked at [the linked `cmake-rs` issue](https://github.com/rust-lang/cmake-rs/issues/240) and couldn't figure out how the optimization flag filtering was restricted to cargo's `debug` profile (that doesn't mean it isn't; I just couldn't see how). I've cloned the dependency and patched it in locally so I can try to see how any changes will affect the flags and ultimately resolve this... 

Thanks again for pointing me in the right direction!

**Comment by vlovich** [comment]
Look at the flag filtering logic I pointed to in cmake-rs and how the build type is part of the variable they apply the filtering to. It's really convoluted and weird.

**Comment by AsbjornOlling** [comment]
> Regardless of if your patch is enabled, and regardless of the cargo / Rust profile (where you're saying it should be there when compiled for release), /O2 is missing for me in the RELEASE flags [...] I suspect this is how it is for Asbjørn too.

> look at the generated CMakeCache.txt and the CXXFLAGS_RELEASE and CFLAGS_RELEASE variables. The /O2 is only missing when Rust is built debug. In --release the /O2 is there.

From what I can tell, I am only getting the /O2 flags when using the default cargo profile (debug?).

<details>
<summary>The /O2 /DNDEBUG and /Ob2 flags are only used when I don't build with --release</summary>

Running in this repo at commit 8b11c5c, so with @vlovich's patch.

With --release:

```
$ cargo clean
$ cargo build --release
$ cat ./target/release/build/llama-cpp-sys-2-9f60549f3cef90cc/out/build/CMakeCache.txt | Select-String CMAKE_CXX_FLAGS_RELEASE:STRING 
CMAKE_CXX_FLAGS_RELEASE:STRING= -nologo -MD -Brepro
$ cat ./target/release/build/llama-cpp-sys-2-9f60549f3cef90cc/out/build/CMakeCache.txt | Select-String CMAKE_C_FLAGS_RELEASE:STRING
CMAKE_C_FLAGS_RELEASE:STRING= -nologo -MD -Brepro
```

Without --release

```
$ cargo clean
$ cargo build
$ cat ./target/release/build/llama-cpp-sys-2-6ebad67f70f390a6/out/build/CMakeCache.txt | Select-String CMAKE_CXX_FLAGS_RELEASE:STRING 
CMAKE_CXX_FLAGS_RELEASE:STRING= /O2 /DNDEBUG /Ob2 -nologo -MD -Brepro
$ cat ./target/release/build/llama-cpp-sys-2-6ebad67f70f390a6/out/build/CMakeCache.txt | Select-String CMAKE_C_FLAGS_RELEASE:STRING
CMAKE_C_FLAGS_RELEASE:STRING=  /O2 /DNDEBUG /Ob2 -nologo -MD -Brepro
```
</details>


If I apply the suggested changes to examples/usage.rs and test with and without --release, I get this:

<details>
<summary>usage.rs benchmarks</summary>

Note that I am using `--features vulkan`. 

This is all running on a Windows 11 virtual machine, hosted by a proxmox hypervisor with an Nvidia A6000 GPU and the Nvidia 572.16 driver.

With --release:

```
$ cargo run --features vulkan --example usage --release -- ./Meta-Llama-3-8B-Instruct-Q5_K_M.gguf
decode 64 tokens in 1.07 s, speed 59.55 t/s
```

Without --release:

```
$ cargo run --features vulkan --example usage -- ./Meta-Llama-3-8B-Instruct-Q5_K_M.gguf
decode 64 tokens in 0.89 s, speed 72.26 t/s
```

For comparison, llama-bench.exe of the pre-built vulkan binaries from llama.cpp github release b4745 gives 83.41 t/s

The drop to 60t/s is significant, but not as bad as I've seen in other tests...

</details>

Changing the conditional in build.rs as @babichjacob suggested removes the difference in benchmarking.
Could we do that? Is there some reason that this improvement should only apply to debug profiles?

I also tried building for x86_64-pc-windows-gnu, to try and workaround the msvc weirdness by using mingw/gcc instead. I haven't been able to make it work yet.

**Comment by vlovich** [comment]
I have no objection. Accelerators (Vulkan/CUDA) generally experience less of a slowdown but I'm curious why there's still a slowdown. I wonder if the FFI boundary UTF8 validation converting into Rust String from CStr is really adding that much overhead
